### PR TITLE
Store notification content as binary to support unicode

### PR DESCRIPTION
**Pull request by darshanasbg**
### Proposed changes in this pull request

$subject. Related to,
* https://github.com/wso2/product-is/issues/21235

### When should this PR be merged

After merging,
- https://github.com/wso2/carbon-identity-framework/pull/6120

### Follow up actions

When updating product-is versions, both the framework change and this should be update together.

### CODE REVIEW

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/dao/AppNotificationTemplateDAO.java [hunk 1]
```diff
@@ -204,12 +203,11 @@ public void updateNotificationTemplate(NotificationTemplate notificationTemplate
         String channelName = notificationTemplate.getNotificationChannel();
 
         NamedJdbcTemplate namedJdbcTemplate = JdbcUtils.getNewNamedJdbcTemplate();
-        try {
+        try (InputStream contentStream = getContentStream(notificationTemplate))  {
```
Comment (inthirakumaaran) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/dao/AppNotificationTemplateDAO.java:206]:
```suggestion
        try (InputStream contentStream = getContentStream(notificationTemplate)) {
```

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/dao/AppNotificationTemplateDAO.java [hunk 2]
```diff
@@ -223,6 +221,8 @@ public void updateNotificationTemplate(NotificationTemplate notificationTemplate
                     String.format("Error while updating %s template %s of type %s from application %s in %s tenant.",
                             channelName, locale, displayName, applicationUuid, tenantId);
             throw new NotificationTemplateManagerServerException(error, e);
+        } catch (IOException e) {
+            throw new NotificationTemplateManagerServerException("Error while closing content stream.", e);
```
Comment (inthirakumaaran) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/dao/AppNotificationTemplateDAO.java:225]: can this error also occur while processing the stream, not just while closing it?

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/util/I18nEmailUtil.java [hunk 3]
```diff
@@ -281,4 +285,32 @@ public static String getNotificationLocale() {
                 ? IdentityUtil.getProperty(I18nMgtConstants.NOTIFICATION_DEFAULT_LOCALE)
                 : I18nMgtConstants.DEFAULT_NOTIFICATION_LOCALE;
     }
+
+    public static InputStream getContentStream(NotificationTemplate notificationTemplate) {
+
+        String[] templateContent = new String[]{notificationTemplate.getSubject(), notificationTemplate.getBody(),
+                notificationTemplate.getFooter()};
+        String content = new Gson().toJson(templateContent);
+        byte[] contentByteArray = content.getBytes(StandardCharsets.UTF_8);
```
Comment (inthirakumaaran) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/util/I18nEmailUtil.java:294]: Are we going to hard code the encoding to UTF_8?

Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/util/I18nEmailUtil.java:294]: Yes.. As its a implementation detail, it should be fine.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/util/I18nEmailUtil.java [hunk 4]
```diff
@@ -281,4 +285,32 @@ public static String getNotificationLocale() {
...
+        try {
+            byte[] contentByteArray = contentStream.readAllBytes();
+            String content = new String(contentByteArray, StandardCharsets.UTF_8);
+            String[] templateContent = new Gson().fromJson(content, String[].class);
+            if (templateContent != null && templateContent.length == 3) {
+                notificationTemplateResult.setSubject(templateContent[0]);
+                notificationTemplateResult.setBody(templateContent[1]);
+                notificationTemplateResult.setFooter(templateContent[2]);
+            } else {
+                throw new SQLException("Invalid content data.");
```
[the remark this answers is not in the document]
Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/util/I18nEmailUtil.java:310]: This validation is already exists during add & update[1][2].

[1] https://github.com/wso2-extensions/identity-event-handler-notification/blob/837271431fdac6df48732fca27666302f24aa961/components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/EmailTemplateManagerImpl.java#L681C9-L707C10
[2] https://github.com/wso2-extensions/identity-event-handler-notification/blob/master/components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java#L649-L676

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/dao/AppNotificationTemplateDAO.java [hunk 5]
```diff
@@ -61,13 +63,12 @@ public void addNotificationTemplate(NotificationTemplate notificationTemplate, S
         String channelName = notificationTemplate.getNotificationChannel();
 
         NamedJdbcTemplate namedJdbcTemplate = JdbcUtils.getNewNamedJdbcTemplate();
-        try {
+        try (InputStream contentStream = getContentStream(notificationTemplate))  {
+            int contentLength = contentStream.available();
```
[the remark this answers is not in the document]
Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/dao/AppNotificationTemplateDAO.java]: Fixed with: https://github.com/wso2-extensions/identity-event-handler-notification/pull/280/commits/4298c197eb3d442b51f6f8689d1b5a9c0cd5af10

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/dao/AppNotificationTemplateDAO.java (hunk 2 above)
Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/dao/AppNotificationTemplateDAO.java]: Fixed with: https://github.com/wso2-extensions/identity-event-handler-notification/pull/280/commits/4298c197eb3d442b51f6f8689d1b5a9c0cd5af10

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/dao/AppNotificationTemplateDAO.java (hunk 1 above)
Comment (darshanasbg) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/dao/AppNotificationTemplateDAO.java]: Fixed with: https://github.com/wso2-extensions/identity-event-handler-notification/pull/280/commits/4298c197eb3d442b51f6f8689d1b5a9c0cd5af10

**Review by inthirakumaaran — APPROVED** [review]

**Review by UdeshAthukorala — APPROVED** [review]

**Pull request merged**